### ISSUE TITLE
Component Updates for 3.0

### ISSUE THREAD

**Issue by claviska** [shoelace-style/webawesome#13]
This issue intends to cover all the fixes, updates, and refactors we need to do [in preparation for 3.0](https://github.com/shoelace-style/shoelace/issues?q=is%3Aopen+is%3Aissue+milestone%3A%223.0+%28future%29%22). Not every item is a breaking change, but anything that _is_ a breaking change should definitely be here.

**Not all components will require updates, but we do need to audit each one to make sure we have a stable API for 3.0.**

Fair game:
- Attributes/Properties
- Methods
- Events
- Slots
- CSS Parts
- Accessibility bugs or concerns
- DX and UX concerns

Out of bounds:

- CSS custom properties (see shoelace-style/shoelace#14)

Feel free to comment below with suggestions or concerns. A task list is provided below so we can check off each component as they get audited. More details for each component are available below the list.

---

### Form Controls

- [ ] Improve framework interop via https://github.com/shoelace-style/shoelace/discussions/1758

### Alert

- [ ] Should alert be split into `<wa-banner>` (inline/static) and `<wa-alert>` (dynamic)? These seem like two different components with different usage and a11y requirements.
- [ ] https://github.com/shoelace-style/shoelace/issues/1359

### Badge

- [ ] We currently add [built-in styles for buttons](https://shoelace.style/components/badge#with-buttons) so badges show up nicely. Should we add similar default styles for any other components?

### Carousel / Carousel Item

- [ ] This is currently experimental and somewhat buggy. It was contributed by a user who still sometimes fixes bugs in it. I don't think we should continue down this path. Maybe we can redesign this one and built it from scratch to make it more stable.

### Color Picker

- [ ] https://github.com/shoelace-style/shoelace/issues/1552
- [ ] Do we want to continue offering the `inline` version? Should it be a separate component?
- [ ] Should the sliders we're using be `<wa-range>` sliders to reduce logic/size?

### Drawer

- [ ] Should we continue to support [`contained`](https://shoelace.style/components/drawer#contained-to-an-element)? This option is weird in that it doesn't make the drawer modal and the drawer doesn't dismiss with <kbd>Esc</kbd>. Never felt right. If we keep it, maybe it can remain modal and containment is strictly for positioning instead.

### Dropdown

- [ ] Can we stop using `<sl-menu>` and `<sl-menu-item>` here and add `<sl-dropdown-item>` to make it more consistent with other components? They can share menu styles if need be, but this API is a remnant from the original dropdown component, which was designed to be a low-level utility. We don't need that now that we have `<sl-popup>` now. See also https://github.com/shoelace-style/shoelace/issues/1688

### Icon

- [ ] Rework to support Font Awesome

### Input

- [ ] Determine if we should split out `<wa-input type="number">` into a separate component or not. Note that doing this would mean other input types should probably follow suit.
- [x] https://github.com/shoelace-style/webawesome/issues/103
- [ ] Do we want to add [input groups](https://getbootstrap.com/docs/4.0/components/input-group/) and, if so, is that a separate component or just slots?

### Menu

- [ ] This component likely needs a rewrite to improve accessibility. As of today, I noticed it's not even reading the labels anymore. VoiceOver just says "menu item" instead of reading the label. https://github.com/shoelace-style/shoelace/issues/1688
- [ ] Add loading state to menu items https://github.com/shoelace-style/shoelace/pull/1700

### Menu Item

- [ ] Add support for `href` and other link attributes (see shoelace-style/shoelace#1351 and #1629)

### Menu Label

- [ ] Remove this and show examples using any heading level and `::slotted()` styles

### Popup

- [ ] Rename to `<sl-popover>` to be more inline with the platform (which was renamed from popup => popover)

### Radio Group

- [ ] Support `disabled` on the radio group controller _as well as_ individual radios
  - https://github.com/shoelace-style/shoelace/discussions/1737

### Radio Button

- [ ] Numerous people have said the selected styles aren't really obvious. Maybe we can improve the design of these a bit.

### Range

- [x] https://github.com/shoelace-style/shoelace/issues/1350

### Select

- [ ] Consider _not_ serializing multiple values with space-delimited options, as it's unintuitive and numerous people have complained. (Not sure what the alternative is here, but consider evaluating how `<select multiple>` submits data and following that pattern.)

### Spinner

- [x] https://github.com/shoelace-style/shoelace/issues/1724

### Tab Group

- [x] https://github.com/shoelace-style/shoelace/issues/1695

### Tag

 - [ ] Should we stop using `<sl-tag>` and style something to look like a tag instead? https://github.com/shoelace-style/shoelace/pull/1741#issuecomment-1828347528

### Tooltip

- [x] https://github.com/shoelace-style/shoelace/issues/1734

### Miscellaneous

- [x] https://github.com/shoelace-style/shoelace/issues/1706
- [x] https://github.com/shoelace-style/shoelace/issues/1718

**Comment by lindsaym-fa** [comment]
> Should alert be split into `<wa-banner>` (inline/static) and `<wa-alert>` (dynamic)? These seem like two different components with different usage and a11y requirements.

Rather than `wa-banner` and `wa-alert`, what if we split off the toast behaviors into something like `wa-toast-stack` or `wa-notifications`?

Assuming I'm on the same page, `wa-banner` and `wa-alert` seem like they'd have a ton of overlap. The primary difference we're trying to account for is live region behavior, right? Using a live region for an inline banner is completely valid (I've seen this often with form validation messages), but it has different requirements compared to a region that could host multiple banners simultaneously, like a toast stack.

Live region requirements vary depending on context, so I'd err on leaving those decisions in the hands of user with some guidance from us. For `wa-alert`, we could expose a subset of values for the native `role=""` attribute. If we take the guidance "No ARIA is better than bad ARIA" to heart, we'd default to leaving the attribute absent and allow users to define `log`, `status`, or `alert`. If `role` is defined, we make sure that's on some element in the DOM, which could be the host `<wa-alert>` element.

If we want to rename `wa-alert` to `wa-banner` to eliminate any confusion surrounding the fact that it wouldn't inherently assume `role="alert"`, that seems like a good idea. We could also consider making the component visible by default and replacing `open` with `hidden`, if that's more fitting.

For toasts, we'd want to have a single live region for the entire stack, not the individual elements within it. We could move all of the toast and toast stack behaviors into a standalone component and have a similarly exposed `role=""` attribute for the parent element. This has a neat benefit of allowing any component to behave as a toast, not just banners/alerts — I could see someone wanting to use cards or even chips/tags for a log of notifications.

**Comment by lindsaym-fa** [comment]
> We currently add [built-in styles for buttons](https://shoelace.style/components/badge#with-buttons) so badges show up nicely. Should we add similar default styles for any other components?

Internal Web Awesome has built-in badge positioning for cards, where the badge can appear at the top left, top center, or top right. That seems super useful.

**Comment by lindsaym-fa** [comment]
> Can we stop using `<sl-menu>` and `<sl-menu-item>` here and add `<sl-dropdown-item>` to make it more consistent with other components?

This is interesting. I see in the open Shoelace issue, you say, "This would effectively eliminate menu and menu item, which people frequently mistake for nav menus." Would `wa-dropdown` cover the trigger element + menu + items? Does `wa-menu` get subsumed and eliminated, and do `wa-menu-item` and `wa-menu-label` become `wa-dropdown-item` and `wa-dropdown-group`, respectively?

Could folks be confusing `wa-menu` for navigation because they're looking for something like a navigation bar and "menu" is the closest thing they find? In that case, the solution might be to add a nav bar component rather than rebranding menus. Overall, "menu" feels more semantically accurate than "dropdown", although I'd happily concede to the masses on that point.

**Comment by lindsaym-fa** [comment]
> Determine if we should split out `<wa-input type="number">` into a separate component or not. Note that doing this would mean other input types should probably follow suit.

We've already split out many input types into their own components — `button`, `checkbox`, `radio`, `color`, `range`, etc. If we can add a tailored layer of semantics on top of the native HTML `input` that improves DX, I'd say it's warranted. I could see something like `wa-text-input` continuing to handle any types that only require plain text presentations, e.g. `email`, `tel`, `search`, `text`, and `url`. The remaining — especially `date` and `time` in addition to `number` — could be improved as standalone components and simplify the existing `wa-input` interface. (`password` is an interesting case that I could see remaining a part of the current component or becoming its own.)

> Do we want to add [input groups](https://getbootstrap.com/docs/4.0/components/input-group/) and, if so, is that a separate component or just slots?

Following the pattern of `wa-button` > `wa-button-group` and `wa-radio` > `wa-radio-group`, I vote for a separate component.

**Comment by kelseythejackson** [comment]
>  Numerous people have said the selected styles aren't really obvious. Maybe we can improve the design of these a bit.

Curious about this one. Those who have said this, did they give any more details as to why?

**Comment by kelseythejackson** [comment]
> Should we continue to support [contained](https://shoelace.style/components/drawer#contained-to-an-element)? This option is weird in that it doesn't make the drawer modal and the drawer doesn't dismiss with Esc. Never felt right. If we keep it, maybe it can remain modal and containment is strictly for positioning instead.

Playing around with it, I think dropping support might be the best route.

**Comment by claviska** [comment]
> Curious about this one. Those who have said this, did they give any more details as to why?

The active button looks a lot like the primary button, which is a reasonable argument.

**Comment by lindsaym-fa** [comment]
> The active button looks a lot like the primary button, which is a reasonable argument.

Maybe we want to validate something like this with our beta group/early adopters?
<img width="336" alt="Screenshot 2023-12-01 at 10 12 38 AM" src="https://github.com/shoelace-style/webawesome/assets/126139086/d9c410db-8c81-4422-8310-0b323ecd0a97">

**Comment by lindsaym-fa** [comment]
Another improvement to consider: Could we support a wider range of components/elements within `wa-radio-group`? I commonly see radio inputs that aren't strictly radios or a unified button group, like these:
<img width="300" alt="Screenshot 2023-12-07 at 1 31 08 PM" src="https://github.com/shoelace-style/webawesome/assets/126139086/687d1b01-92ea-4573-825b-e01978a55b7b">
<img width="300" alt="Screenshot 2023-12-07 at 1 32 35 PM" src="https://github.com/shoelace-style/webawesome/assets/126139086/bbaa2b4b-4802-4054-aa7e-53a21d55b464">
<img width="300" alt="Screenshot 2023-12-07 at 1 33 32 PM" src="https://github.com/shoelace-style/webawesome/assets/126139086/7f4904c9-3313-40cf-96ee-2cbf862bbd7d">

**Comment by KonnorRogers** [comment]
## ElementInternals: This should be a huge win for forms, so +1 from me. There are some limitations and accessibility gaps to sort out, but having a standardized approach should help.

## Alert

Banners and alerts tend to behave differently for accessibility.

`role="banner"` vs `role="alert"`.

I also feel like from a styling perspective, alerts and banners have different styles. Our current alerts have a "border" on the top and have a border radius, both of which you wouldn't expect on a banner that sits at the top of a section and tends to take up the entire width of it's container. 

Example of a "banner" alert with no custom styling today:

Perhaps this a good case for a "variant", since they will share a lot of similarities for things like being able to dismiss, close, etc.

https://codepen.io/paramagicdev/pen/MWLdWbJ?editors=1100

Maybe this is a good use-case for a `"variant"`?

## Carousel

This is a tough one, because our existing implementation does "work". I'm also curious if we should look at any existing carousel libraries and see if there's something we can wrap. I'm not sure if the carousel is that buggy and that's why we get so many reports, or because everyone is using it.

## Color Picker

Reusing existing components is always nice, and makes theming more "consistent", it does require exposing nested `exportparts`, but I think it's fine.

## Drawer

Contained makes sense to keep for styling, I guess the bigger question is if it should be dialog-like and if that should be a separate attribute, especially in applications where the side-navigation flows in underneath the header.

## Dropdown

I'm struggling to think of a situation where you would have a dropdown, with items, that wouldn't benefit from the accessibility of the menu with keyboard support. Would the dropdown item do anything for keyboards? Or just be a pre-styled element in the dropdown with no keyboard semantics?

## Input

I like that we generally try to follow the HTML equivalent, and don't particularly love the idea of splitting off into a separate component, but `<input>` is really overloaded in regular HTML, so it makes sense to split some of the types off so we don't balloon the API and all the permutations. I'm on the fence and don't have any strong feelings either way.

## Menu Item

+1 for hrefs! It seems we have more and more interactive components that are "button"-like / "link"-like, so it'll be important to sort out how we can share logic across similar components.

## Popup -> Popover 👍 

## Radio Group

Disabled is cool! I like being able to disable individual / entire groups.

## Range

+1 to css properties, obviously we can't cover every permutation, and people would need to know the browser prefixes to do more custom styling, but its a good start.

I'd also love to add vertical sliders! (Yes, I know they're a pain, but they're nice to have)


## Radio Group 

- Radio Group: @lindsaym-fa With any luck, as we transition to ElementInternals, these more "grouped" form controls like "radio-button" wouldn't necessarily need to live inside of a `<radio-group>` and would function much closer to the platform. The `<radio-group>` is more like a nice `<fieldset>` wrapper.

## Select 

> Select:  Consider not serializing multiple values with space-delimited options, as it's unintuitive and numerous people have complained. (Not sure what the alternative is here, but consider evaluating how `<select multiple>` submits data and following that pattern.)

the platform puts everything into an array on the form data based on the `name` of the `<select>`

https://codepen.io/paramagicdev/pen/abXxrym

As for multi-select, it feels a lot like a "search" filter right now. Obviously the default browser implementation for multi-select is much more like a "listbox".

## Tags

I think tags are cool as is. Although they look a lot like badges currently 🤷‍♂️

**Comment by claviska** [comment]
> I'm struggling to think of a situation where you would have a dropdown, with items, that wouldn't benefit from the accessibility of the menu with keyboard support. Would the dropdown item do anything for keyboards? Or just be a pre-styled element in the dropdown with no keyboard semantics?

It would still have keyboard support, just less boilerplate HTML to create it.

**Comment by KonnorRogers** [comment]
> > I'm struggling to think of a situation where you would have a dropdown, with items, that wouldn't benefit from the accessibility of the menu with keyboard support. Would the dropdown item do anything for keyboards? Or just be a pre-styled element in the dropdown with no keyboard semantics?
> 
> It would still have keyboard support, just less boilerplate HTML to create it.

So essentially a `<menu>` with a built in `popover`?

**Comment by claviska** [comment]
This was getting way to hard to follow, so I've broken each item out into discussions. Please follow the conversation there now:

https://github.com/shoelace-style/webawesome/discussions?discussions_q=is%3Aopen+label%3A3.0
